Feature/ti 2471 season editor form version

bartosz-bujakowski opened the pull request:
Proposal for to-do list ordering:

- [x] MVP
  - [x] Add sticky header
  - [x] Select subCompetences in testItems (possibly with text instead of ID) by name and automatically fill competence ids
  - [x] Ensure ids are unique for testItem and feedbackQuestion (or disable the field and generate it automatically) (as discussed today).
- [ ] After MVP or "if quick to do"
  - [ ] Description fields as textarea
  - [ ] Add popup for delete confirmation
  - [ ] Add statistics (number of episodes, number of test items, number of test items per episode etc.)
  - [ ] ~Hide log form button under some dev tools icon~
  - [ ] Add periodic save state and go back to previous state
  - [ ] Pretty load file button
  - [ ] Editable test item IDs.
- [ ] Park/hold
  - [ ] Test item support for toolType and search
  - [ ] Check for missing properties - (one known key missing is: seasonEndMessage)

eickler commented on apps/season-editor/src/pages/components/sub-competences.tsx:
```diff
@@ -0,0 +1,87 @@
+                disabled
+                type="text"
+                {...register(
+                  `subCompetences.${index}.competenceId` as "subCompetences.0.subCompetenceId"
```
I think there's a sub missing here ... it also prints the wrong ID when I load the canary season.yaml.

<img width="493" alt="Screenshot 2023-11-27 at 12 56 18" src="https://github.com/talent-digital/talentkit/assets/797483/f60f762a-9d38-4043-8d64-841205cabcb6">

bartosz-bujakowski replied:
outdated

ryanoshea-td commented:
[c] Simple and ease to use 👍

[m] Definitely a quick UI win (as @eickler mentioned) is have the focus on `Name` when we add one the different competence levels. 

[m] The set initial id needs a button click to send. Couldn't that be done with enter `enter` rather than tabbing to the button? 

[m] Also, when appropriate (also as André says), something so that the user can find the relevant competence and episode without scrolling back up. The would make me happy as a user.

[c] I think the background color should be #83adb2

[c] I see inconsistency in the layout.
- The competence areas are contained in one panel 
- But individual episodes, test items and feedback questions are in separate
- [c] I much prefer the way you handle it in competences 

[c] Not sure about needing a separate panel for the `add some item` buttons. Integrating all the UI for each section into the same panel would remove the need for that button panel.

Happy to hear your feedback to my comments!

bartosz-bujakowski commented:
I've implemented changes from the comments with the exception of the layout changes (let's wait and see what Gina's feedback will be).

I've created a task for each of the "to-do" items

eickler reviewed (changes requested):
Some findings from a round of user interface testing: 

[m] Can we preserve the order in which the data is shown across saves? For example, when I created test items, they were created as test-150--1, --2, --3. But when I reloaded the season.yaml, they were shown as test-150--3, --1, --2. That will probably be irritating to users.
[m] I can delete an episode that is still used in a test item. It doesn't cause an error or prevent the user from saving, which is good. But there is no error shown and the drop-down will be empty on reload (= dangling reference in the yaml).
[c] If there is not principal issue, I would make the test item IDs editable after the MVP. It helps quite a lot to have readable test item IDs during analysis
[c] Why are there double dashes in test item IDs?

[q] We could in principle simply make this available through the /app Netlify deployment mechanism, couldn't we? That would also give us access to saving the data straight into a tenants season endpoint, wouldn't it? I guess we would need to check how the endpoint deals with references and stale data, e.g.., deleted episodes/competences/... and provide a mechanism for that.
[q] (to be tried) Can we actually use links to our "S3 CMS" e.g. in the image links?

[q] Can you collect the "post MVP" improvements into another user story on the backlog? We can use that story as a discussion vehicle after this one is closed and after Gina/Roman/... had a look.

bartosz-bujakowski commented:
@eickler 
> [m] I can delete an episode that is still used in a test item. It doesn't cause an error or prevent the user from saving, which is good. But there is no error shown and the drop-down will be empty on reload (= dangling reference in the yaml).

Good catch, forgot about this, fixed.

> [c] If there is not principal issue, I would make the test item IDs editable after the MVP. It helps quite a lot to have readable test item IDs during analysis

https://talentdigital.atlassian.net/jira/software/projects/TI/boards/10?selectedIssue=TI-2497

> [c] Why are there double dashes in test item IDs?

That's technical implementation leaking out, I use it to get the number and increment it, could as well be a single dash but a double dash is easier :) Thought it makes no difference since it's an Id

> [q] We could in principle simply make this available through the /app Netlify deployment mechanism, couldn't we? That would also give us access to saving the data straight into a tenants season endpoint, wouldn't it? I guess we would need to check how the endpoint deals with references and stale data, e.g.., deleted episodes/competences/... and provide a mechanism for that.

I guess we could. Seems like a different user story. This would also make our deploy scripts kind of obsolete.

> [q] (to be tried) Can we actually use links to our "S3 CMS" e.g. in the image links?

To have a preview of the image or to be able to select from there?

> Let's collect the "post MVP" improvements into another user story on the backlog. We can use that story as a discussion vehicle after this one is closed and after Gina/Roman/... had a look.

👍

eickler commented:
> I guess we could. Seems like a different user story. This would also make our deploy scripts kind of obsolete.

I think the "Github way" comes from several requirements:

* Version control (though you can deploy to several tenants to get "poor man's version control" a la Webflow).
* Store all additional artefacts including format configurations.
* Store software artefacts for formats (unrelated to YAML deployment).

> > [q] (to be tried) Can we actually use links to our "S3 CMS" e.g. in the image links?
> 
> To have a preview of the image or to be able to select from there?

Yeah, to paste here a link like /S3/images/episode1preview.svg or something.
